src/llmcompressor/pipelines/sequential/helpers.py

```diff
@@ -0,0 +1,365 @@
+
+    :param graph: subgraph of model graph
+    :param input_names: argument names of the compiled forward function
+    :param consumed_names: argument names which are not used by any subsequent subgraphs
+        and can therefore be deleted from the intermediates cache
+    """
+
+    graph: Graph
+    input_names: Set[str]
+    consumed_names: Set[str]
+
+    def compile_forward(self) -> Callable[[Any], Any]:
+        """
+        Generate and compile code for executing this subgraph
+
+        :return: function which, when called, executes this subgraph
+        """
+        code = self.graph.python_code("self")
+        exec(code.src, code.globals)
+        return code.globals.get("forward")
+
+
+def trace_subgraphs(
+    model: PreTrainedModel,
+    sample_input: Dict[str, Any],
+    sequential_targets: List[str],
+    ignore: List[str],
+) -> List[Subgraph]:
+    """
+    Trace a model to produce subgraphs, where each sequential target belongs to exactly
+    one subgraph and where executing each subgraph in order is equivalent to executing
+    the original model
+
+    :param model: model being traced
+    :param sample_input: inputs whose values will change during execution but whose
+        __len__, __bool__, and __contains__ values are assumed constant across batches
+    :param sequential_targets: list of patterns matching sequential targets
+    :param ignore: list of patterns matching modules to ignore during tracing
+    :return: a list of Subgraphs in order of execution
+    """
+    # find modules
+    sequential_targets = match_modules(model, sequential_targets)
+    ignore = match_modules(model, ignore)
+
+    # initialize arguments
+    tracer = get_tracer(model, sequential_targets, ignore)
+    concrete_args = populate_concrete_args(model, sample_input)
+
+    # trace
+    with (
+        calibration_forward_context(model),
+        HooksMixin.disable_hooks(),
+    ):
+        graph = GraphModule(
+            model,
+            tracer.trace(
+                model,
+                dummy_inputs=sample_input,
+                concrete_args=concrete_args,
+                complete_concrete_args_with_inputs_not_in_dummy_inputs=False,
+                # bug in trace throws an error for variadic
+                # args and kwargs in function signature
+            ),
+        )
+
+    # copy metadata
+    graph.config = model.config
+    graph.class_for_deserialization = model.__class__
+    graph.device = model.device
+
+    # perform subgraph partition
+    partitions = topological_partition(graph, sequential_targets)
+    subgraphs = partition_graph(model, partitions)
+    trace_consumed_names(subgraphs)
+
+    return subgraphs
+
+
+def get_tracer(
+    model: Module, sequential_targets: Set[Module], ignore: Set[Module]
+) -> HFTracer:
+    """
+    Get a tracer specialized for the given model. The resulting tracer will not trace
+    inside of sequential targets, ignored targets, or offloaded modules.
```

Comment:
Trying to understand this comment. If the resulting tracer does not trace offloaded modules, how does this work for cases when we have parts of the model offloaded?

Reply:
What does the tracer actually trace?

Reply:
> Tracing within sequential targets and ignored targets is unnecessary, and tracing within offloaded modules may result in meta tensors being added to the model graph

When a module is "not traced", this means that the internals of module are not traced, but the module still appears in the graph as a `call_module` node.

Reply:
For example, in this model graph, the internals of `language_model_model_layers_39` are not traced (we don't see individual nodes for the attention mechanism, ffn, ect.) but all the operations after the module are traced with high granularity (we see individual `pow`, `.to` and `.dtype` functions).

<img width="529" alt="Screenshot 2025-01-06 at 23 22 15" src="https://github.com/user-attachments/assets/ab7295e5-6040-49c0-a6dd-f63e658ebe59" />

Reply:
As a side note, even if a module contains untraceable code internally, if the internals of the module skip tracing via `ignore` or `sequential_targets` or `has_offloaded_params`, then the model graph as a whole will still be traceable, but just with less granularity.

Reply:
@mgoin The tracer traces all of the objects and operations that are need perform a forward pass of the model.